pennylane/hf/tapering.py

```diff
@@ -0,0 +1,204 @@
+        if len(term.wires) == 1:
+            ops = [ops]
+        for op, wire in zip(ops, wires):
+            if op in ["PauliX", "PauliY"]:
+                E[idx][wire + num_qubits] = 1
+            if op in ["PauliZ", "PauliY"]:
+                E[idx][wire] = 1
+
+    return E
+
+
+def _reduced_row_echelon(binary_matrix):
+    """ Returns the reduced row echelon form (RREF) of a matrix in a binary finite field Z_2 """
+
+    binary_matrix = deepcopy(binary_matrix)
+    shape = binary_matrix.shape
+
+    for irow, icol in zip(range(shape[0]), range(shape[1])):
+
+        # find value and index of largest element in remainder of column icol
+        krow = irow + np.argmax(binary_matrix[irow:, icol])
+
+        # swap rows krow and irow
+        binary_matrix[krow], binary_matrix[irow] = deepcopy(binary_matrix[irow]), deepcopy(binary_matrix[krow])
+
+        # store remainder columns of the row irow
+        pvtcols = binary_matrix[irow, icol:]
+
+        # get the column icol and set its irow element to 0 to avoid XORing pivot row with itself
+        currcol = deepcopy(binary_matrix[:, icol])
+        currcol[irow] = 0
+        binary_matrix[:, icol:] ^= np.outer(currcol, pvtcols)
+
+    return binary_matrix.astype(int)
+
+
+def _kernel(binary_matrix):
+    """ Computes the kernel of a binary matrix on the binary finite field Z_2 """
+
+    # Get the columns with and without pivots
+    pivots = (binary_matrix.T!=0).argmax(axis=0)
+    nonpivots = np.setdiff1d(range(len(binary_matrix[0])), pivots)
+
+    # Initialize the nullspace
+    null_vector = np.zeros((binary_matrix.shape[1], len(nonpivots)), dtype=int)
+    null_vector[nonpivots, np.arange(len(nonpivots))] = 1
+
+    # Fill up the nullspace vectors from the binary matrix
+    null_vector_indices = np.ix_(pivots, np.arange(len(nonpivots)))
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices]%2
+
+    return null_vector.T
+
+
+def generate_taus(nullspace, num_qubits):
+    """Generate generators tau from the nullspace
+
+    Args:
+        nullspace (list): kernel of the binary matrix corresponding to the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+
+    Returns:
+        generators (list): list of generators of symmetries, taus, for the Hamiltonian.
+
+    """
+
+    generators = []
+
+    for null_vector in nullspace:
+        tau = qml.Identity(0)
+        for idx, op in enumerate(
+            zip(null_vector[:num_qubits], null_vector[num_qubits:])
+        ):
+            x, z = op
+            if x == 0 and z == 0:
+                tau @= qml.Identity(idx)
+            elif x == 1 and z == 0:
+                tau @= qml.PauliX(idx)
+            elif x == 1 and z == 1:
+                tau @= qml.PauliY(idx)
+            else:
+                tau @= qml.PauliZ(idx)
+        ham = qml.Hamiltonian([1.0], [tau], simplify=True)
```

Comment:
Why are we using `simplify=True` here?

Reply:
To remove the Identities operations in the observables. For example: `[I0 Z0 Z1 I2 I3]` becomes `[Z0 Z1]`. Should I remove it?

Reply:
I think it should be OK, if the generators are correct.